kong/runloop/handler.lua

```diff
@@ -1438,7 +1438,7 @@ return {
 
       local upstream_status_header = constants.HEADERS.UPSTREAM_STATUS
       if kong.configuration.enabled_headers[upstream_status_header] then
-        local upstream_status = ctx.buffered_status or tonumber(sub(var.upstream_status or "", -3))
+        local upstream_status = ctx.buffered_status or tonumber(sub(var.upstream_status or "", -3)) or ngx.status
```

Comment:
I'm wondering if it would be more appropriate to set a special literal like `nil` rather than setting it to `ngx.status`. After all, in this case, the request doesn't actually reach the upstream.

Reply:
@ms2008 If the response is returned from upstream, we set the value of this header with either `ctx.buffered_status` (in case it is a buffered response) or `var.upstream_status` (respect the status from upstream). In the case that the response is a cached one, we have to respect the status that's already cached. In the logic of `proxy-cache` plugin, the `ngx.status` is set with last cached status. Thus, I suppose here it should consider `ngx.status` as the last alternative.

`nil` is not a suitable one, as it is not a valid value for a header. And `0` is not suitable as well, as it doesn't coincide with the cached response.

Reply:
Hmm, that sounds reasonable too. But I think in this scenario, we should clearly define the return value of these `X-Kong-Upstream-*` response headers, otherwise it will be confusing for the user. https://github.com/Kong/kong/blob/ed0b96de49da824d76122cbaf7fc15060405afe5/kong.conf.default#L921-L925

Reply:
Let me revise our doc and describe this more clearly: This **header** always keep same with the status cached by proxy-cache plugin when requests hit the cache.